### ISSUE TITLE
Problem about GPU memory

### ISSUE THREAD

**Issue by lxy443626128** [NVIDIA/DALI#2598]
my train task is a video classification.
my data format is: [ frame1 frame2 frame3 ...  frame8 label] 
and the image size is at least 1080p.
my dali dataloader:

`
class ExternalInputIterator(object):

    def __init__(self, images_dir, batch_size):
        self.images_dir = images_dir
        self.batch_size = batch_size
        self.num_segments = 8
        with open(self.images_dir, 'r') as f:
            self.files = [line.rstrip() for line in f if line is not '']
        shuffle(self.files)

    def __iter__(self):
        self.i = 0
        self.n = len(self.files)
        return self

    def __len__(self):
        return len(self.files)

    def __next__(self):
        frame_1  = []
        frame_2  = []
        frame_3  = []
        frame_4  = []
        frame_5  = []
        frame_6  = []
        frame_7  = []
        frame_8  = []
        labels = []

        for _ in range(self.batch_size):
            f1, f2, f3, f4, f5, f6, f7, f8 ,label = self.files[self.i].split(' ')

            f = open(f1, 'rb')
            img1 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_1.append(img1)
            
            f = open(f2, 'rb')
            img2 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_2.append(img2)

            f = open(f3, 'rb')
            img3 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_3.append(img3)

            f = open(f4, 'rb')
            img4 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_4.append(img4)

            f = open(f5, 'rb')
            img5 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_5.append(img5)

            f = open(f6, 'rb')
            img6 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_6.append(img6)

            f = open(f7, 'rb')
            img7 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_7.append(img7)

            f = open(f8, 'rb')
            img8 = np.frombuffer(f.read(), dtype = np.uint8)
            frame_8.append(img8)
            
            labels.append(np.array([label], dtype = np.float32))

            self.i = (self.i + 1) % self.n

        return (frame_1, frame_2, frame_3, frame_4, frame_5, frame_6, frame_7, frame_8, labels)
`

`class ExternalTrainPipeline(Pipeline):

    def __init__(self, batch_size, eii, num_threads, device_id):
        super(ExternalTrainPipeline, self).__init__(batch_size, num_threads, device_id ,seed=12)

        self.source = ops.ExternalSource(source = eii, num_outputs = 9)
        self.decode = ops.ImageDecoder(device = "mixed", output_type = types.RGB)
        self.resize = ops.RandomResizedCrop(
            device="gpu",
            size=[256, 256],
            interp_type=types.INTERP_LINEAR,
            random_aspect_ratio=[0.75, 4.0 / 3.0],
            random_area=[0.08, 1.0],
            num_attempts=100,
        )
        self.cmnp = ops.CropMirrorNormalize(device="gpu",
                                            output_layout=types.NCHW,
                                            crop=(224, 224),
                                            mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
                                            std=[0.229 * 255, 0.224 * 255, 0.225 * 255])
        self.coin = ops.CoinFlip(probability=0.5)
    def define_graph(self):

        rng = self.coin()

        f1, f2, f3, f4, f5, f6, f7, f8 ,label = self.source()
        img1 = self.decode(f1) 
        img2 = self.decode(f2)
        img3 = self.decode(f3)
        img4 = self.decode(f4)
        img5 = self.decode(f5)
        img6 = self.decode(f6)
        img7 = self.decode(f7)
        img8 = self.decode(f8)
        
        img1 = self.resize(img1)
        img2 = self.resize(img2)
        img3 = self.resize(img3)
        img4 = self.resize(img4)
        img5 = self.resize(img5)
        img6 = self.resize(img6)
        img7 = self.resize(img7)
        img8 = self.resize(img8)
        
        img1 = self.cmnp(img1, mirror=rng)
        img2 = self.cmnp(img2, mirror=rng)
        img3 = self.cmnp(img3, mirror=rng)
        img4 = self.cmnp(img4, mirror=rng)
        img5 = self.cmnp(img5, mirror=rng)
        img6 = self.cmnp(img6, mirror=rng)
        img7 = self.cmnp(img7, mirror=rng)
        img8 = self.cmnp(img8, mirror=rng)
        
        return (img1, img2, img3, img4, img5, img6, img7, img8, label)`

`class DALIWrapper(object):

    def gen_wrapper(dalipipeline):
        for data in dalipipeline:
            img1 = data[0]['img1'].unsqueeze(1)
            img2 = data[0]['img2'].unsqueeze(1)
            img3 = data[0]['img3'].unsqueeze(1)
            img4 = data[0]['img4'].unsqueeze(1)
            img5 = data[0]['img5'].unsqueeze(1)
            img6 = data[0]['img6'].unsqueeze(1)
            img7 = data[0]['img7'].unsqueeze(1)
            img8 = data[0]['img8'].unsqueeze(1)
            label = data[0]['label']
            images = torch.cat([img1, img2, img3, img4, img5, img6, img7, img8], dim=1)
            #images = images.view(-1,3,224,224)
            yield images, label
        dalipipeline.reset()

    def __init__(self, dalipipeline):
        self.dalipipeline = dalipipeline

    def __iter__(self):
        return DALIWrapper.gen_wrapper(
            self.dalipipeline
        )`

when I run:

`    eii = ExternalInputIterator(train_data, args.batch_size)

      pipe = ExternalTrainPipeline(batch_size=args.batch_size, eii=eii, num_threads=5, device_id = 0)
      pipe.build()
      train_loader = DALIGenericIterator(pipe,output_map= ["img1","img2","img3","img4","img5","img6","img7","img8", "label"], size=len(eii))
      lenTrain = len(train_loader)
      train_loader = DALIWrapper(train_loader)`

the gpu 0 memory is basically full！！！

**Comment by lxy443626128** [comment]
@JanuszL

**Comment by JanuszL** [comment]
Hi,
This is somehow expected. You can read more about DALI memory management in[ this section of the documentation](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/advanced_topics.html#memory-consumption).
What you can do is:
- lower the batch size
- replace ImageDecoder + RandomResizedCrop by ImageDecoderRandomCrop + Resize
